core/wasm-interface/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+	const VALUE_TYPE: ValueType = ValueType::I32;
+	fn into_value(self) -> Value { Value::I32(self.ptr as _) }
+}
+
+impl<T> TryFromValue for Pointer<T> {
+	fn try_from_value(val: Value) -> Option<Self> {
+		match val {
+			Value::I32(val) => Some(Self::new(val as _)),
+			_ => None,
+		}
+	}
+}
+
+/// The word size used in wasm. Normally known as `usize` in Rust.
+pub type WordSize = u32;
+
+/// The Signature of a function
+#[derive(Eq, PartialEq, Debug)]
+pub struct Signature {
+	/// The arguments of a function.
+	pub args: Cow<'static, [ValueType]>,
+	/// The optional return value of a function.
+	pub return_value: Option<ValueType>,
+}
+
+impl Signature {
+	/// Create a new instance of `Signature`.
+	pub fn new<T: Into<Cow<'static, [ValueType]>>>(args: T, return_value: Option<ValueType>) -> Self {
+		Self {
+			args: args.into(),
+			return_value,
+		}
+	}
+
+	/// Create a new instance of `Signature` with the given `args` and without any return value.
+	pub fn new_with_args<T: Into<Cow<'static, [ValueType]>>>(args: T) -> Self {
+		Self {
+			args: args.into(),
+			return_value: None,
+		}
+	}
+
+}
+
+/// A reference to a host function.
+#[derive(Eq, PartialEq, Debug)]
+pub struct FunctionRef {
+	/// The signature of the function.
+	pub signature: Signature,
+	/// The index of the function at the host.
+	pub index: usize,
+}
+
+impl FunctionRef {
+	/// Create a new instance of `FunctionRef`.
+	pub fn new(signature: Signature, index: usize) -> Self {
+		Self {
+			signature,
+			index,
+		}
+	}
+
+	/// Add the given offset to the internal index.
+	pub fn offset_index(&mut self, offset: usize) {
+		self.index += offset;
+	}
+}
+
+/// Context used by `Externals` to interact with the allocator and the memory of the wasm instance.
+pub trait ExternalsContext {
+	/// Read memory from `address` into a vector.
+	fn read_memory(&self, address: Pointer<u8>, size: WordSize) -> Result<Vec<u8>, String> {
+		let mut vec = Vec::with_capacity(size as usize);
+		unsafe { vec.set_len(size as usize); }
+		self.read_memory_into(address, &mut vec)?;
+		Ok(vec)
+	}
+	/// Read memory into the given `dest` buffer from `address`.
+	fn read_memory_into(&self, address: Pointer<u8>, dest: &mut [u8]) -> Result<(), String>;
+	/// Write the given data at `address` into the memory.
+	fn write_memory(&mut self, address: Pointer<u8>, data: &[u8]) -> Result<(), String>;
+	/// Allocate a memory instance of `size` bytes.
+	fn allocate_memory(&mut self, size: WordSize) -> Result<Pointer<u8>, String>;
+	/// Deallocate a given memory instance.
+	fn deallocate_memory(&mut self, ptr: Pointer<u8>) -> Result<(), String>;
+}
+
+/// Something that provides implementations for host functions.
+pub trait Externals {
+	/// Try to resolve the function with the given name and signature.
+	fn resolve_function(name: &str, signature: &Signature) -> Result<Option<FunctionRef>, String>;
+
+	/// Returns the number of host functions this type provides.
+	fn function_count() -> usize;
```

Comment:
Why would we need the count?

Reply:
Currently we only have these 2 externals. When we come to the point that we have `N` externals. It will be used to call into the correct `execute_function`.